### ISSUE TITLE
Deadlock when unsibscribe from channels

### ISSUE THREAD

**Issue by adamvas** [pusher/pusher-websocket-dotnet#111]
Hi, 
It can casue deadlock if you are unsubscribing from a channel in an async awaited method, like:
```
public override Task Shutdown()
{
	channel.UnbindAll();
	channel.Unsubscribe(); // <- wrapped **blocking** async call  : Task.WaitAll(_pusher.ChannelUnsubscribeAsync(Name)); 
	return pusherClient.DisconnectAsync();
}
```
eg:
```
public async Task DeactivatedAsync()
{
    ...
    await connector.Shutdown(); // <- async **non-blocking** call
    ...
}
```

Thats bcs of this:
```
 public void Unsubscribe()
{
    Task.WaitAll(_pusher.ChannelUnsubscribeAsync(Name)); 
}
```

_Blocking async call in a non-blocking async call._

Best practice is usually **async all-way-down**.

**Comment by erwin42** [comment]
I will look at fixing this. However you could use the following instead:
```cs
public override Task Shutdown()
{
	channel.UnbindAll();
	pusherClient.UnsubscribeAsync(channel.Name);
	return pusherClient.DisconnectAsync();
}
```

Please let me know if the above works?

**Comment by adamvas** [comment]
Hi @erwin42 , sure I changed my code to this already. It works as expected, just wanted to highlight it.
Is there any reason behind that `Unsubscribe` is not async by default, I mean `Task` instead of `void`?

**Comment by erwin42** [comment]
We would have removed the function, but we kept it for backwards compatibility. You will notice no mention of the `Channel.Unsubscribe` method in the Readme doc. We are trying to discourage people from using the method in favour of `Pusher.UnsubscribeAsync`.

I tried to repro your issue but couldn’t. I will leave a test in for it in our next release.

**Comment by erwin42** [comment]
I will add `public async Task UnsubscribeAsync()` to the `Channel' class in our next release to help with this issue

**Comment by elverkilde** [comment]
hi @adamvas and thanks again for reporting an issue. The async method was added in #112 and released in version `2.1.0`.
